pkg/app/pipedv1/livestatereporter/livestatereporter.go

```diff
@@ -45,127 +51,225 @@ type Reporter interface {
+	snapshotTicker := time.NewTicker(r.snapshotFlushInterval)
+	defer snapshotTicker.Stop()
 
-	for _, reporter := range r.reporters {
-		// Avoid starting all reporters at the same time to reduce the API call burst.
-		time.Sleep(10 * time.Second)
+	for {
+		select {
+		case <-snapshotTicker.C:
+			r.flushSnapshots(ctx)
 
-		group.Go(func() error {
-			reporter.logger.Info("starting app live state reporter for plugin")
-			return reporter.Run(ctx)
-		})
+		case <-ctx.Done():
+			r.logger.Info("app live state reporter has been stopped")
+			return nil
+		}
 	}
+}
 
-	r.logger.Info(fmt.Sprintf("all live state reporters of %d plugins have been started", len(r.reporters)))
+func (r *reporter) flushSnapshots(ctx context.Context) {
+	r.logger.Info("start flushing snapshots")
 
-	if err := group.Wait(); err != nil {
-		r.logger.Error("failed while running", zap.Error(err))
-		return err
+	r.logger.Info("updating git repositories")
+	for id, cfgRepo := range r.pipedConfig.GetRepositoryMap() {
+		if _, ok := r.repoMap[id]; ok {
+			r.logger.Info("the repo is cloned, so pulling repository", zap.String("repo-id", id))
+			err := r.repoMap[id].Pull(ctx, cfgRepo.Branch)
+			if err != nil {
+				r.logger.Error("failed to pull repository", zap.String("repo-id", id), zap.Error(err))
+			}
+			continue
+		}
+
+		r.logger.Info("the repo is not cloned, so cloning repository", zap.String("repo-id", id))
+		repo, err := r.gitClient.Clone(ctx, id, cfgRepo.Remote, cfgRepo.Branch, fmt.Sprintf("%s/%s", r.workingDir, id))
+		if err != nil {
+			r.logger.Error("failed to clone repository", zap.String("repo-id", id), zap.Error(err))
+			continue
+		}
+		r.repoMap[id] = repo
 	}
 
-	r.logger.Info(fmt.Sprintf("all live state reporters of %d plugins have been stopped", len(r.reporters)))
-	return nil
-}
+	apps := r.appLister.List()
 
-type pluginReporter struct {
-	pluginName            string
-	snapshotFlushInterval time.Duration
-	appLister             applicationLister
-	apiClient             apiClient
-	pluginClient          pluginapi.PluginClient
-	logger                *zap.Logger
-}
+	// TODO: Set the limit based on the piped config
+	limit := runtime.GOMAXPROCS(0) / 2
 
-func (pr *pluginReporter) Run(ctx context.Context) error {
-	pr.logger.Info("start running app live state reporter", zap.Duration("snapshot-flush-interval", pr.snapshotFlushInterval))
+	appsPerReporter := len(apps) / limit
+	if appsPerReporter == 0 {
+		appsPerReporter = 1
+	}
 
-	snapshotTicker := time.NewTicker(pr.snapshotFlushInterval)
-	defer snapshotTicker.Stop()
+	appsGrouped := make([][]*model.Application, 0)
+	for i := 0; i < len(apps); i += appsPerReporter {
+		end := i + appsPerReporter
+		if end > len(apps) {
+			end = len(apps)
+		}
+		appsGrouped = append(appsGrouped, apps[i:end])
+	}
 
-	for {
-		select {
-		case <-snapshotTicker.C:
-			pr.flushSnapshots(ctx)
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.SetLimit(limit)
```

Comment:
We use the `appsGrouped` to limit concurrency, so we don't need this SetLimit.
Or, we can use only `eg.SetLimit` and simply call `eg.Go` with all apps.

Reply:
Thanks, you're right. Also there is no reason to use errgroup, so I fixed to use wait group instead of it.
[af57c01](https://github.com/pipe-cd/pipecd/pull/5457/commits/af57c01d1718c38eb416f2a590ec2f386d99e632)